[Blueprint] Introduce map and abstract algebra module for new function framework

mcpl-sympy opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes  #18028

#### Brief description of what is fixed or changed

`map` module and `algebras.abstract` module are introduced.

#### Other comments

Abstract algebra is introduced with strict mathematical background. Every structures, i.e. Group, Field, Vectorspace, ... are implemented with adequate operators. This allows one to construct a function space, which in turn allows the operation between functions such as `(f+g)(x)` or `(f*g)(x)`.  
Dual space and homomorphism can be introduced with these tools. I expect that this modules can supplant current `Add`, `Mul` and `Pow` as well.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

- map
    - map module is introduced.

- algebras
    - algebras.abstract module is introduced.

<!-- END RELEASE NOTES -->

JSS95 committed:
Make map package and module

Co-authored-by: sylee957 <sylee957@gmail.com>

JSS95 committed:
Define parameters and add test

JSS95 committed:
Add Map and AppliedMap to test_args

mcpl-sympy commented:
@sylee957 I am trying to take on the new feature that you suggested, and I would be grateful if you supervise this PR.

JSS95 committed:
Remove unused import from test_map

JSS95 committed:
Update setup.py

JSS95 committed:
Add nargs logic

JSS95 committed:
Add str printing of Map and AppliedMap

JSS95 committed:
Add latex printing of Map and AppliedMap

JSS95 committed:
Add pretty printing of Map and AppliedMap

JSS95 committed:
Delete unnecessary import in map.py

JSS95 committed:
Add tests for printing of Map without parameter

mcpl-sympy committed:
Merge branch 'master' into 18028_Map

JSS95 committed:
Add doctest to Map and AppliedMap

JSS95 committed:
Add "sympy.map.tests" to setup.py

sylee957 commented on sympy/map/map.py:
```diff
@@ -0,0 +1,193 @@
+        return self.args[3]
+
+    @classmethod
+    def _find_nargs(cls):
```
I thought that the number of arguments can be matched to the number of cartesian products of the domain

oscarbenjamin replied:
Probably worth looking at how `Lambda/ImageSet` handles this including for nested tuples. A multi-arg Lambda only works when multiple base sets are given to ImageSet. Otherwise ProductSets can be unpacked using a nested tuple signature:
```julia
In [9]: A = FiniteSet(1, 2)                                                                                            

In [10]: B = FiniteSet(3, 4)                                                                                           

In [11]: ImageSet(Lambda((x, y), x+y), A, B)                                                                           
Out[11]: {x + y | x ∊ {1, 2} , y ∊ {3, 4}}

In [12]: ImageSet(Lambda(((x, y),), x+y), A * B)                                                                       
Out[12]: {x + y | (x, y) ∊ {1, 2} × {3, 4}}
```

mcpl-sympy replied:
I applied it in commit `d43072a`. How is it now?

JSS95 committed:
Add InverseMap class and its tests

JSS95 committed:
Change nargs logic to use ProductSet

Also, simplified doctests and tests of map.py

JSS95 committed:
Add latex, str, pretty printing of InverseMap

JSS95 committed:
Add more docstring to InverseMap

JSS95 committed:
Add test_AppliedMap

JSS95 committed:
Add IdentityMap

JSS95 committed:
Add printing to IdentityMap

JSS95 committed:
Add doctest to IdentityMap

JSS95 committed:
Improve code quality

JSS95 committed:
Add 1-arg case in IdentityMap's docstring

sylee957 commented:
Although we can have a more generic definition of cartesian product, I don't think that programmatic concept of variadic length arguments means a cartesian product `S.Reals**n`, and it can introduce some confusion whether tuples of any length like `(1,), (1, 2, 3), (1, 2, 3, ...)` should belong `S.Reals**n`.

Is it totally necessary to support variadic length of arguments?
I don't think that `Map` may be extended for the scope of this, while `Function` could be, because of the lack of mathematical tools.

mcpl-sympy commented:
@sylee957 I believe that you intended to comment in #19754. We may need `S.Reals**n` someday since it is commonly encountered in mathematics. Just as SymPy supports `Derivative(f(x), (x,n))`, there might be some way to support this as well.  
However, I'd say we don't need this feature right now.

JSS95 committed:
Introduce CompositeMap and its test

JSS95 committed:
Add printing for CompositeMap

sylee957 commented:
I'd also have some questions about how should `parameters` be used for.
I suppose that you had intended to include variations and branches as `parameters`, however, I don't think that they are always necessary because you can always denote them as multivariable functions. 

Maybe you have intended to handle them separately because of the issues of defining inverse functions for some rectangular mappings, but I think that `parameters` should also link to set-theoritic domains.

mcpl-sympy commented:
@sylee957
I referred to the wikipedia page: [https://en.m.wikipedia.org/wiki/Parameter#Mathematical_functions]

JSS95 committed:
Add doctest to CompositeMap

JSS95 committed:
Change IdentityMap to be unary

JSS95 committed:
Introduce UndefinedMap class

Now, Map is abstract base class.

Also, printing of classes are changed slightly and
sstrrepr printing of maps are fixed.

JSS95 committed:
Improve code quality

JSS95 committed:
Define domain and codomain of CompositeMap

Fix flake8 test failure as well

sylee957 commented:
I'm still not getting that `parameters` need such special treatment than multivariate function, and it doesn't need the definition of domains even if we usually expect them to belong to integers or reals or complexes.

JSS95 committed:
Delete parameters feature

JSS95 committed:
Fix flake8 test failure

JSS95 committed:
Enhance tests of map module

JSS95 committed:
Enhance composition and add Q.invertible

Q.invertible for maps are added.
CompositeMap now checks domains and codomains.
InverseMap now checks invertibility.
Tests are updated accordingly.

JSS95 committed:
Update docstring of CompositeMap

JSS95 committed:
Improve code quality

mcpl-sympy commented:
@sylee957 @oscarbenjamin   
I deleted `parameters` feature and made `Q.invertible` comaptible with map classes. Can you review this, please?

oscarbenjamin commented:
This looks nice as an abstract way of working with functions. It isn't clear to me how this would work if some significant classes like `exp` or `log` were migrated to this scheme. I suspect that is the more difficult part of making something like this work.

Maybe a new pull request could be opened on top of this that shows how a couple of functions could be migrated. Particular problems are the old assumptions, where to put the rewrite methods etc. Probably this would change the meaning of some things like `replace`, if `exp` becomes an instance of Basic rather than a subclass.

JSS95 committed:
Define printing precedence for CompositeMap

JSS95 committed:
Make map.rst and let classes imported in sympy.__init__

JSS95 committed:
Add __matmul__ to return evaluated CompositeMap

JSS95 committed:
Introduce CompositionalMapPow class

Basic tests and printings are introduced. Actual evaluatation logic is
not implemented yet.

JSS95 committed:
Use name "IteratedMap" instead of "CompositionalMapPow"

JSS95 committed:
Modify logic of composition evaluation

Composition with identity map or inverse map are not handled in
_eval_composite now. Rather, they are handled in CompositeMap.flatten

JSS95 committed:
Simplify the construction logic of classes

JSS95 committed:
Make IteratedMap handle 0, 1, and negative iteration

JSS95 committed:
Add test for _eval_iteration

JSS95 committed:
Define evaluation of IteratedMap

JSS95 committed:
Improve code quality

JSS95 committed:
Introduce as_base_iternum method

Nested IteratedMap evaluation is enabled.

TODO: establish interface between CompostieMap and IteratedMap.
Consecutively composited maps will be evaluated to IteratedMap.

mcpl-sympy commented:
@oscarbenjamin 
Scalar function cannot be implemented right away, since #19759 is needed for rigorous definition of function spaces (which will allow  features like `sin**2`).
I will introduce binary operator in this PR, and make function composition dependent on that. Then, I'll combine `Set` and `Map` to introduce mathematical structures in next PR. Next, binary operator `+`, `-`, etc will be introduced as Basic instance. Only after then it will be possible to start migrating existing functions into this framework.

oscarbenjamin commented:
I'm not sure it makes sense to merge PRs building up an abstract function framework if it isn't yet clear that it can work for the important cases that we already have. It's possible that attempting to migrate some important functions over to this will lead to significant changes in the design.

JSS95 committed:
Introduce BinaryOperator class

JSS95 committed:
Change binary.py to operator.py

JSS95 committed:
Connect CompositeMap and IteratedMap

JSS95 committed:
Add RestrictedMap

JSS95 committed:
Add printing of RestrictedMap

mcpl-sympy commented:
Then let me tinker on some more and notice you when it's ready.

JSS95 committed:
Allow defined inverse cancellation of composite maps

JSS95 committed:
Modify printing of Maps

Now, domains and codomains are printed.

JSS95 committed:
Update docstrings

JSS95 committed:
Introduce UndefinedSet class

JSS95 committed:
Modify tests and doctests

This fixes test failures and add docstrings.

JSS95 committed:
Add intersection and union dispatch of UndefinedSet

JSS95 committed:
Introduce SetElement class

Add `element` and `_element` method to Set

JSS95 committed:
Add more test to test_UndefinedSet

These tests checks the relation between
UndefinedSet and fancy sets.

JSS95 committed:
Add __all__ to undefinedset.py

JSS95 committed:
Modify sets.rst

JSS95 committed:
Add support for nullary map (domain={()})

JSS95 committed:
Use name "arity" instead of "nargs" on Map

JSS95 committed:
Fix dispatchers for UndefinedSet

Some methods were resulting infinite loop.

JSS95 committed:
Modify Set.contains and Set.__contains__

Now, these method check `_contained` method of the element before
checking `self._contains`.

JSS95 committed:
Make Map.is_restriction check arity

JSS95 committed:
Introduce algebras/abstract module and AlgebraicStructure class

JSS95 committed:
Fix doctest failure

JSS95 committed:
Change AlgebraicStructure.operators

This property now only returns self.args[2]

JSS95 committed:
Make AlgebraicStructure.sets and operators Tuple

Order of these matters, since it will be used to distinguish the sets
and operators in complex structures (ring, module, etc)

JSS95 committed:
Add codomain checking to AppliedMap

JSS95 committed:
Introduce Magma class

JSS95 committed:
Fix doctest failure

JSS95 committed:
Enhance docstring and test of map.py

Fix trivial typos in docstrings of structure.py and undefinedset.py

JSS95 committed:
Make ProductSet._contains checks its arguments' contains, not _contains

JSS95 committed:
Introduce subclasses of AppliedMap

AppliedBinaryOperator and AppliedAssociativeOperator
are introduced, with adequate doctests, tests,
and printings.
printing.precedence.py is updated with
AppliedBinaryOperator as well.

JSS95 committed:
Enhance test_composite.py

JSS95 committed:
Define evaluation of identity map iteration

JSS95 committed:
Implement Semigroup class

JSS95 committed:
Implement codomain checking of AppliedMap

JSS95 committed:
Introduce element method to AlgebraicStructure

JSS95 committed:
Implement identities to BinaryOperator

JSS95 committed:
Implement Monoid class to group.py

JSS95 committed:
Modify Map printing

Now, attributes such as latex_name are referred to.

JSS95 committed:
Introduce Q.associative predicate

Removed AssociativeOperator and its applied class.

JSS95 committed:
Introduce LeftDivision and RightDivision to operator.py

Added Q.left_divisible and Q.right_divisible predicate.

JSS95 committed:
Add domain and codomain checking logic to BinaryOperator

JSS95 committed:
Implement Quasigroup

JSS95 committed:
Remove _eval_(left/right)_division

LeftDivision and RightDivision itself does not
contruct other object any more. This means
that rather then making '-' as evaluated
division operator of '+', '-' itself should be
subclass of division operator class.

JSS95 committed:
Split group.py into several modules

With this change, structures such as left/right
quasigroups or inverse semigroup will be implemented.

JSS95 committed:
Implement LeftQuasigroup and RightQuasigroup

Now, Quasigroup is subclass of these classes.

JSS95 committed:
Add Map to Q.commutative

JSS95 committed:
Introduce InverseOperator and ExponentOperator

Printing is not implemented yet

JSS95 committed:
Add InverseElement and ExponeneElement

Printing for those are implemented.
Todo: Modify as_base_exp method to be specified with operator

JSS95 committed:
Make as_base_exp take argument

This argument is the instance of Map. This is necessary because various
"exponentiation" operator can exist. For example, sin**2 can be interpreted
as sin composite with sin, or sin multiplied with sin. If it is composite,
then decomposing to base and exp with respect to multiplication should
return (sin**2, 1).

JSS95 committed:
Implement Loop, Group, and AbelianGroup classes

JSS95 committed:
Move group-like classes to abstract.group package

Ring-like structures and algebraic modules will be introduced in
other packages.

JSS95 committed:
Introduce map/add.py and map/mul.py

These modules implement classes for general
addition and multiplication.

This PR implements only skeletal features. A lot
more needs to be done afterwards.

JSS95 committed:
Introduce ring module

Todo: Implement collect and distribute method
to AdditionOperator and MultiplicationOperator

JSS95 committed:
Introduce isapplied function

JSS95 committed:
Introduce distribute, undistribute and as_coeff_Mul methods

JSS95 committed:
Introduce Field class

Also connect distribute and undistribute methods
to Ring.

Operations still need tons of work, but they will
be done in the long term.

JSS95 committed:
Introduce Module and VectorSpace

JSS95 committed:
Introduce FunctionSet class

Todo: introduce FunctionAdd and FunctionMul

JSS95 committed:
Add pre-defined structures to S

S.Integers_ring, S.Reals_field, S.Complexes_field are added.

JSS95 committed:
Modify inverse-element logic

For group operators, inverse of element returns ExponentElement.
scalar_divide is added accordingly.

JSS95 committed:
Enhance scalar_mul and derived operators

JSS95 committed:
Fix doctest failure

JSS95 committed:
Add Map.apply

This method is an API for map and its applied class.
Changed the classes in operator.py, add.py and mul.py to use this method. This
change made AppliedBinaryOperator needless, so it is deleted.
Also, relation between InverseElement and ExponentElement is modified. Now,
if the base operator is associative, ExponentElement is used instead of
InverseElement.

JSS95 committed:
Fix test failures

JSS95 committed:
Delete functionspace.py

FunctionSet is moved to composite.py

sylee957 commented on sympy/sets/undefinedset.py:
```diff
@@ -0,0 +1,122 @@
+    name : str
+        Name of the undefined set.
+
+    supersets : set of Sets, optional
```
I have thought that this also suffers from similar problem as the previous `CoordSystem` relations, that objects need to be defined first in order to define full relation.
Would there be any cases where some mutual relation can occur?

mcpl-sympy replied:
In `CoordSystem`, mutual reference was crucial since SymPy cannot deduce the inverse transformation between curvilinear coordinates. But this superset-subset relation is simple, so mutual relation won't be needed, I suppose.

sylee957 replied:
Actually, I start to find that mathematicians often denote equality between sets as A is subset of B and B is subset of A.
I think that it’s be better to allow full relation diagram here, for this reason

mcpl-sympy replied:
Sounds reasonable. Instead of defining relational diagram, how about introducing assumption predicate something like ``Q.subset(A, B)``?

sylee957 replied:
That can also be a valid idea

mcpl-sympy replied:
My group is planning to implement a finitary relation class, involving equality and inequality, as a subclass of ``ProductSet``. After that is done and API with assumption module is established,

JSS95 committed:
Modify composite.py to use BinaryOperator

Precedence and printing are modified as well.

JSS95 committed:
Change isapplied to isappliedmap

JSS95 committed:
Modify the attributes of group-like structures

JSS95 committed:
Enhance ring operations

Abstract addition and abstract multiplication is supported.
Subtraction, division and power is defined for ring.
Printing of operators is enhanced.

JSS95 committed:
Add latex printing of ring of int and fields of reals and complexes

JSS95 committed:
Enhance docstring and test of ring and field

JSS95 committed:
Enhance algebraic module class

Introduced VectorAddition and ScalarMultiplication
classes.

Todo: fortify the testings

JSS95 committed:
Fix bugs and add test to module.py

JSS95 committed:
Write more docstring to VectorAddition

JSS95 committed:
Enhance VectorSpace class

JSS95 committed:
Introduce ConstantMap

JSS95 committed:
Fix typo

JSS95 committed:
Implement algebraic algebra

VectorMultiplicationOperator is introduced.

JSS95 committed:
Introduce functionspace.py

JSS95 committed:
Introduce FunctionVectorMultiplicationOperator

JSS95 committed:
Enhance doctest and fix error

JSS95 committed:
Simplify the docstrings in functionspace.py

JSS95 committed:
Allow __add__ and etc. to Map

JSS95 committed:
Make commutative map sort the arguments

JSS95 committed:
Apply commutability to mul.py

JSS95 committed:
Introduce range to Map

JSS95 committed:
Add domain and codomain to IteratedMap

JSS95 committed:
Fix pretty printing of AppliedMap

JSS95 committed:
Enhance structure finding logic

Fixed test failures

JSS95 committed:
Remote 'is_' from properties

Properties such as is_commutative are related with metaclass and is
not really expandable. Instead, map module uses .commutative and
.associative properties.

JSS95 committed:
Introduce FunctionExponent

JSS95 committed:
Fix test failures

JSS95 committed:
Introduce trigonometric functions

JSS95 committed:
Introduce ReciprocalFunction

JSS95 committed:
Introduce reciprocal trigonometric functions

mcpl-sympy committed:
Merge branch 'master' into 18028_Map

JSS95 committed:
Improve code quality

JSS95 committed:
Improve flake8 quality

JSS95 committed:
Build document rst for new modules

JSS95 committed:
Update setup.py

oscarbenjamin commented:
@mcpl-sympy this PR adds almost 10k lines of code and there has been hardly any discussion of the different parts of it.

It isn't clear to me what this does altogether although I can see various bits that look useful.

I recommend breaking pieces out of this into separate PRs because I really don't think we can merge this as is. It is too big to possibly review and contains all kinds of separate pieces that would individually require some discussion.

I just want to repeat that I don't think we should build a new function framework unless we are sure that it can replace the old one. It's hard to see everything that is changed here but it doesn't look like this PR changes any of the existing functions and I don't see anything that addresses the key obstacles to replacing the old function framework. In particular we need to decide how to handle the old assumptions and how to handle things like `isinstance(exp(x), exp)` (which is used in operations like replace etc).

Performance is also a major concern. If this PR hasn't actually replaced any of the existing classes then we don't have a way to benchmark it. If this ties in things like the sets module as part of basic function evaluation then I suspect that might slow things down a lot.

JSS95 committed:
Fix flake8 error

mcpl-sympy commented:
@oscarbenjamin 
I finished with this just now and have been writing a summary. To be short, this PR introduces functions as `Basic` instance and function space. Most part of the codes is related to the latter one.

The reason why I implemented function space is to make things like `sin + cos` possible with strict mathematical background (My previous attempts to do this with current `Add`, `Mul` and `Pow` failed terribly).

I will do some benchmarks and report here.

oscarbenjamin commented:
> make things like `sin + cos` possible

This has already been discussed and some contributors are of the opinion that we should not make things like `sin + cos` possible. That's one reason that it is important to discuss/agree basic design ideas before implementing too much.

oscarbenjamin commented:
I think that the changes to sets could be made as a PR on their own since it looks like they don't depend on any of the other changes.

oscarbenjamin commented:
This looks like such a major change that I think it needs to be explained in detail somewhere with a rationale justifying the design decisions.

There should also be some significant rst docs documenting the whole idea (not just a collection of docstrings).

Why do we need all of these classes like Rings etc? Are they separately useful?

What problems are solved by all of this?

There are many open issues that this either solves or should be solving but only one is listed in the OP.

mcpl-sympy commented:
Then the docstring in core/functions mentioning sin+cos would better be removed.  
Besides, once we have functions as Basic, implementing sin+cos will be too enticing to be ignored. Whole concepts like dual space, topology, etc can be introduced rigorously with this.

mcpl-sympy commented:
I tried some tests with new addition function. With cache enabled, new add was slower than old add when small number of arguments are passed. However, there wasn't much difference when >1000 arguments are passed. There are some rooms for optimization so I'll improve it later.

sylee957 commented:
As discussed in #19832, I think that this project can be a good direction to build a framework to extendable algebra and functions without modifying the core, and disintegrate much of stuff like `Matrix`, `Vector`, `Quaternion`, `Boolean` or such outside of the old function framework.

But most of the old function frameworks symbolic integrations, limit evaluations, series evaluations or such doesn't need abstractions, so I think that even if things doesn't work out effectively or migration is difficult, those old function framework can operate in isolated way and this can be implemented in parallel way.

oscarbenjamin commented:
I don't disagree but the purpose of all this code needs to be explained and these changes need to be reviewed properly.

I've looked through all of the discussion in this PR and in the linked issue (there is not much discussion) and I've looked through the 10000 lines of code in the diff here and I really don't understand why we need all of this. I'm not saying we don't need need it: I'm saying that I don't understand because there's no explanation anywhere.

Where did this all come from?

One paragraph responses are not going to be sufficient as an explanation of what this means and why we should include it in the codebase.

I suggest this:

1. Anything that can be split out of this PR as a separate PR should be split out so that it can reviewed as a smaller piece of work.
2. Write up an explanation of the design that is implemented here. Explain what this is for, why we should have it, why it should be implemented in this way, what we should do with it in future, how users can use it etc.

If this is solving some problems then that's great but right now I don't understand what problems it solves or how. I don't think that it's a good idea to build a large parallel framework without a clear plan (understood by all) for what we will do with it.

sylee957 commented:
We may need the Ring classes because it can be used to keep arg-invariance for Poly

mcpl-sympy commented:
@oscarbenjamin 
I understand. I was running short of time due to my circumstance and had to materialize what I had in my mind quickly - so I decided to make commit first and chisel down later.

This PR can be split down to 3 parts - sets, maps, and algebras.

## Set

`UndefinedSet` and `SetElement` are introduced to set module. Also, `_contained` method is introduced for `Basic` so that an element can define which set it belongs to.  
This change is incidental to implementing domain and codomain to map object - it does not solve anything right now. It can be separated to another PR, or be abandoned safely.

## Map

Basic implementation of the ideas from #18028 and #4787 are done in `map/map.py`. What it solves and what can be done in the future are explained in these issues. `map/composite.py` includes function composition operator and composite function.

At the first time, I introduced composite function as a class. However a question struck me; *isn't function composition is binary operator? If I'm introducing function as object, why not introduce function composition as operator as well?* That lead to `map/operator.py`, which contains basic binary operator. Composite operator in `map/composite.py` subclass that class.

Introducing composite function also led to function iteration, which requires function power. Now, since I need mul-like operator and pow-like operator, there is no reason that I should avert from implementing addition, multiplication and power operator. This led to `map/add.py` and `map/mul.py`.

## Algebra

Having abstract sets and abstract functions means that I can establish algebraic structures. My first intention was to make `Ring` class as subclass of `Basic`. Currently, domains in `poly` module are not subclass of `Basic`. This leads to unorthodox behavior such as:
```python
>>> expr = Poly(x, x, domain=QQ)
>>> expr.func(*expr.args) == expr
False
```
It clearly needs to be fixed.

Also, I wanted to introduce operation between functions i.e. `sin + cos`. The fact that functions can be operated requires that same operation can be done between functions' images. This can be defined by introducing vector space object, but I think we can just introduce `FunctionAdd`, `FunctionMul` and `FunctionPow` as subclass of `Add`, `Mul`, and `Pow`.

So this needs to be discussed first: do we need to define operation between functions? If we don't, then most of the contents in `algebras/abstract` can be removed.

sylee957 commented:
The most difficult part of using binary operator is that sympy had been doing additions, multiplications in n-ary operations than binary operations.

I find that the formalism for n-ary operators, variadic functions, algebras over rectangular matrices which are common practices for programming, are also limitations of abstract algebra, that may be resolved in the category theory.
However, I think that the operator as object can still be useful as some signature even if it can't be used practically, such that we don't have to end up every classes of integer ring, modular integer ring, ... where the only differences are the operator itself.

oscarbenjamin commented:
The two issues you have referenced also have hardly any discussion. Those are about creating a new function framework but this has thousands of lines of algebraic structures that seem very much orthogonal to me.

I still don't understand what all of these operators and so on are for. What is a user going to do with them?

Yes it's true that adding functions gives something like a vector space but if we want to allow adding functions then we probably also want to allow multiplying them and composing them. At that point a vector space is not a sufficient representation of the properties. What use does a VectorSpace class have? What would anyone do with that class directly?

Note that we do already have various implementations of groups, rings etc around the codebase. Is this intended to supplant those?

Is the idea that this PR will provide the Basic instances that correspond to the Poly domains? If so does it do that?

It sounds like the algebra part is implemented on top of the functions part so those can be separate PRs although one would be built on top of the other.

I think it will be much easier to review and understand this work and make sure that it gets merged if it is broken down as much as possible into separate pieces.

mcpl-sympy commented:
@sylee957  Associative binary operators are designed to be able to take n arguments, so that would not be problem.

mcpl-sympy commented:
@oscarbenjamin 
>What use does a VectorSpace class have? What would anyone do with that class directly?

Multiplication between functions is multiplication between vectors, which is an algebra over a field. It is implemented as `Algebra` object.

`VectorSpace` is a wraps scalar field and vector group.
Suppose this situation:
1. F is a field of real numbers
2. G is a group of 2\*2 matrices.
3. a, b are X -> F functions.
4. c, d are X -> G functions.

Here, 2\*a, a\*b, a\*c are commutative. But c\*d is not commutative 

For this case, I Vector space V can be introduced, whose scalar field is F and vector group is G.
Next, a set of function S is introduced. Codomain of any element of S is subset of V. Finally, vector-vector multiplication operator × : S\*S -> S is defined.

When c × b is run, it asks its codomain (S) which is vector function and which is scalar function. In turn, S asks V, and V recognizes b's codomain F and c's codomain G. Thus, × finds out that b and c are commutable so it returns b×c, instead of c×b.

mcpl-sympy commented:
But again, I can abort this abstract algebra logic and combine `Map` with `Add` and `Mul` if we conclude that this design is unnecessary. Consider this PR as just an experiment of the ideas I had in mind.

Before splitting down this PR to pieces, I would like to hear your opinion on function operation (apart from how it will be implemented). My hunch is that we need it - we need `sin + cos`, `exp + tan`, `d/dx + d/dy` or anything else. How do you think of it?

oscarbenjamin commented:
> if we conclude that this design is unnecessary

To be clear: I am not saying that it is unnecessary. Rather it is *unexpected* (to me at least).

I want to understand what this is trying to do and I think that will be much easier if it is broken down into smaller pieces.

mcpl-sympy commented:
Got it. I'll extract Map-related part and make as separate PR.

oscarbenjamin commented:
> Multiplication between functions is multiplication between vectors, which is an algebra over a field. It is implemented as `Algebra` object.

This is only true for linear functions though. Multiplication of nonlinear functions would not give an algebra. Also if we have both multiplication and composition then it is something more than an algebra because we have multiple multiplication operators

mcpl-sympy commented:
I may be wrong, but isn't that a condition for the function to be vector space homomorphism? I thought `(sin*cos)(x) = sin(x)*cos(x)` can be algebraically defined, although they are not homomorphism.

oscarbenjamin commented:
Oh, yes, you're right. It's composition of nonlinear functions that would break bilinearity
